### ISSUE TITLE
Request: Making `import-outside-top-level (PLC0415)` configurable per import section

### ISSUE THREAD

**Issue by Avasam** [astral-sh/ruff#18728]
### Summary

pywin32 has a ton of imports that are not top-level. A lot of those are known to have side-effects (yeah ik, that's a bad pattern, but it's also one of the oldest library out there, not much I can do about it nowadays :P ). Others are just known to be heavy imports. I can't just blindly move everything to a top-level (currently ~400 hits) without heavy scrutiny of what every import does and their context. Even then I'd still end up with tons of `noqa`.

But I still think I can get value out of `import-outside-top-level (PLC0415)` even in legacy projects, for example I wouldn't have any issue just moving all stdlib to top-level.

Rather than making that super specific request, I think it can be generalized to work with https://docs.astral.sh/ruff/settings/#lint_isort_sections which would allow users to get `import-outside-top-level (PLC0415)` to work for their specific needs.

**Comment by ntBre** [comment]
That's an interesting suggestion. Do you think that would be preferable to a more specific config option like `allowed-local-imports`? If I'm following, it sounds like it would end up being more configuration to define the sections and then add imports to them.

I could definitely see having some configuration here, in any case. The rule docs mention that there are valid cases for module-level imports, so it makes sense to me to have an escape hatch other than noqa for those valid cases.

We also have [banned-module-level-imports (TID253)](https://docs.astral.sh/ruff/rules/banned-module-level-imports/#banned-module-level-imports-tid253) for imports that should always be delayed. PLC0415 already respects this list, so this would be a new option for rules that are only sometimes allowed to be local.

**Comment by Avasam** [comment]
`allowed-local-imports` would work for my needs, I think I'd basically copy the list I already have of known side-effect modules for pycln https://github.com/mhammond/pywin32/blob/main/pycln.toml and add any "heavy" or "optional" module as necessary.

<details>
<summary>List of current 30 local modules in pywin32, deduplicated</summary

```py
import argparse
import document_object
import getopt
import importlib.machinery
import os
import platform
import profile
import pythoncom
import pywin.scintilla.find
import regutil
import shutil
import subprocess
import sys
import win32api
import win32com
import win32com.client.connect
import win32com.client.dynamic
import win32com.client.gencache
import win32com.demos.connect
import win32com.server.policy
import win32com.server.register
import win32com.server.util
import win32com.servers.dictionary
import win32com.test.Generated4Test.msword8
import win32con
import win32event
import win32process
import win32ui
import winerror
import winreg
```

</details>

> We also have [banned-module-level-imports (TID253)](https://docs.astral.sh/ruff/rules/banned-module-level-imports/#banned-module-level-imports-tid253) for imports that should always be delayed. PLC0415 already respects this list

Nice, I didn't know. Not a use-case for me, just an example I thought of the top of my head to justify adding configs to `PLC0415`. Maybe the `PLC0415` doc page should link to [banned-module-level-imports](https://docs.astral.sh/ruff/settings/#lint_flake8-tidy-imports_banned-module-level-imports) if it's affected by it? (looks like a simple documentation PR to add an Options section, I can take care of that).

---

So anything in `banned-module-level-imports` *has* to be local
Anything not in either  `banned-module-level-imports` or `allowed-local-imports` *has* to be top-level
Anything in `allowed-local-imports` but not in `banned-module-level-imports` could be either.

**Comment by ntBre** [comment]
Yeah I was thinking the same about the PLC0415 docs. Happy to accept a PR! We may want to squeeze in a mention of/link to the related rule, as well as the setting.

> So anything in banned-module-level-imports has to be local
> Anything not in either banned-module-level-imports or allowed-local-imports has to be top-level
> Anything in allowed-local-imports but not in banned-module-level-imports could be either.

The logic/interaction of the settings does look a little frightening when you put it that way! I think that covers all of the cases, though.

I'm curious if anyone else has other ideas here too.

**Comment by MichaReiser** [comment]
It seems the main culprit here are imports with side effects. Would it make more sense to have a setting that allows marking imports that have side effects rather than having so many global/local settings? 

Marking an import (or module) as having side effect would mean that ruff shouldn't move that import, ever.

**Comment by Avasam** [comment]
I think that completely independent of this request, having a way to identify side-effect imports would be beneficial (for [unused-import (F401)](https://docs.astral.sh/ruff/rules/unused-import/#unused-import-f401) and maybe other rules as well?). Whether that's done through a manual list for now, or with more automated detection later with ty (https://github.com/astral-sh/ruff/issues/8149). Also see discussion in https://github.com/astral-sh/ruff/issues/6601 which seems very related in its needs.
